blog/2023-02-02-good-cognitive-friction.md

```diff
@@ -0,0 +1,264 @@
+> wise embrace history"
+
+A language-level guarantee that state cannot change offers opportunities for caching, runtime
+optimizations and lock-free concurrency. Those attributes are very useful in distributed systems.
+
+## Immutable by default
+
+This is why, similarly to other modern programming languages such as Rust and Go, we are designing
+Wing to be *immutable by default*.
+
+Let's look at an example:
+
+```js
+let my_array = [1,2,3,4];
+```
+
+The above code defines an *immutable* array with the contents `[1,2,3,4]` and assigns it to
+`my_array`. Immutability means that the contents of the object cannot be modified.
+
+So if we try to add an item:
+
+```js
+my_array.push(5);
+//       ^^^^ Unknown symbol "push"
+```
+
+This is because the type of `my_array` is `Array<num>`, which represents an immutable array, it
+simply doesn't have any methods that will cause it to change. In Wing, the following types are
+immutable: `str`, `num`, `bool`, `Array<T>`, `Set<T>` and `Map<T>`.
+
+If I wanted to define it as a mutable array, I will need to be explicit:
+
+```js
+let my_mut_array = MutArray<str>["hello", "world"];
+```
+
+And now we can go wild:
+
+```js
+my_mut_array.push("go wild!"); // OK!
+```
+
+Similarly, we can define other mutable collection types:
+
+```js
+let my_set = MutSet<str>{"hello", "world"};
+let my_map = MutMap<bool>{"dog": true, "cat": false};
+```
+
+> By the way: we are still debating if the standard types should be pascal-cased (e.g. `Array<T>`,
+> `MutArray<T>`) or snake (`array<T>`, `mut_array<T>`). Let us know what you think!
+
+Yes! We are going to make this *slightly* harder to define mutable collections. 
+
+> In the *future*, maybe we will introduce some syntactic sugar like:
+> 
+> ```js
+> let x = mut [1,2,3]; // <-- not a doctor
+> ```
+
+This design concept is what's called "good cognitive friction" (or "mechanical sympathy"). It is
+introduced intentionally in order to make sure the user understands the system better and encourage
+best practices.
+
+## Reassignability
+
+But immutability is not enough! Since we reference our array through `my_array`, the compiler also
+needs to guarantee that `my_array` will always point to the same object.
+
+Let's look at a hypothetic example:
+
+```js
+let i = 10;
+new cloud.Function(inflight () => { print(i); }) as "f1";
+i = 20;
+new cloud.Function(inflight () => { i = i + 9; }) as "f2";
+i = i - 90;
+```
+
+What value will the cloud function print? We can't tell because `i` is reassigned in multiple
+locations and there is absolutely no way to determine its value.
+
+This is where **reassignability** comes into play. In fact, in Wing, the above example would have
+failed compilation:
+
+```js
+   i = 20;
+// ^ variable i is not reassignable
+```
+
+OK, now we can relax. The Wing compiler tells us that `i` is not reassignable.
+
+Reassignability is a form of immutability (it is mutating the reference) and most modern programming
```

Comment:
Was this supposed to be "form of **mutability**" instead of "form of immutability"?

Reply:
But of course